File: vrpc/riot/region.py
from __future__ import annotations

import logging
import re

from ..constants import valorant_log_path
from .local_auth import LocalAuth

logger = logging.getLogger(__name__)

_REGION_TO_SHARD = {
    "na": "na", "latam": "na", "br": "na",
    "eu": "eu", "ap": "ap", "kr": "kr", "pbe": "pbe",
}

_GLZ_RE = re.compile(r"glz-([\w-]+?)-1\.([\w-]+)\.a\.pvp\.net")
_PD_RE = re.compile(r"https://pd\.([\w-]+)\.a\.pvp\.net")


def detect_region_shard(auth: LocalAuth) -> tuple[str, str]:
    try:
        log = valorant_log_path()
        if log.exists():
            txt = log.read_text(encoding="utf-8", errors="ignore")
            m = _GLZ_RE.search(txt)
            if m:
                return m.group(1), m.group(2)
            mpd = _PD_RE.search(txt)
            if mpd:
                shard = mpd.group(1)
                return shard, shard
    except Exception as e:
        logger.debug("Region from log failed: %s", e)

    try:
        r = auth.local_get("/chat/v1/session")
        if r.status_code == 200:
            region = (r.json().get("region") or "").lower().replace("1", "")
            if region:
                return region, _REGION_TO_SHARD.get(region, region)
    except Exception as e:
        logger.debug("Region from chat failed: %s", e)

    logger.warning("Region detection failed, defaulting to 'eu'")
    return "eu", "eu"


def detect_locale(auth: LocalAuth) -> str:
    try:
        r = auth.local_get("/riotclient/region-locale")
        if r.status_code == 200:
            locale = (r.json().get("locale") or "").lower()
            if locale.startswith("tr"):
                return "tr"
    except Exception as e:
        logger.debug("Locale detection failed: %s", e)
    return "en"


def detect_player(auth: LocalAuth) -> tuple[str, str]:
    try:
        r = auth.local_get("/chat/v1/session")
        if r.status_code == 200:
            data = r.json()
            return data.get("game_name", ""), data.get("game_tag", "")
    except Exception as e:
        logger.debug("Player detection failed: %s", e)
    return "", ""

File: electron/riot/region.ts
import { existsSync, readFileSync } from 'fs'
import type { Language } from '@shared/types'
import { valorantLogPath } from '../constants'
import type { LocalAuth } from './localAuth'

const REGION_TO_SHARD: Record<string, string> = {
  na: 'na',
  latam: 'na',
  br: 'na',
  eu: 'eu',
  ap: 'ap',
  kr: 'kr',
  pbe: 'pbe'
}

const GLZ_RE = /glz-([\w-]+?)-1\.([\w-]+)\.a\.pvp\.net/
const PD_RE = /https:\/\/pd\.([\w-]+)\.a\.pvp\.net/


export async function detectRegionShard(auth: LocalAuth): Promise<[string, string]> {
  try {
    const path = valorantLogPath()
    if (existsSync(path)) {
      const txt = readFileSync(path, 'utf-8')
      const m = txt.match(GLZ_RE)
      if (m) return [m[1], m[2]]
      const mpd = txt.match(PD_RE)
      if (mpd) {
        const shard = mpd[1]
        return [shard, shard]
      }
    }
  } catch (e) {
    console.debug('[region] log parse failed:', e)
  }

  try {
    const r = await auth.localGet('/chat/v1/session')
    if (r.status === 200) {
      const region = String(r.data.region || '')
        .toLowerCase()
        .replace('1', '')
      if (region) return [region, REGION_TO_SHARD[region] ?? region]
    }
  } catch (e) {
    console.debug('[region] chat session failed:', e)
  }

  console.warn('[region] detection failed, defaulting to eu')
  return ['eu', 'eu']
}

export async function detectLocale(auth: LocalAuth): Promise<Language> {
  try {
    const r = await auth.localGet('/riotclient/region-locale')
    if (r.status === 200) {
      const locale = String(r.data.locale || '').toLowerCase()
      if (locale.startsWith('tr')) return 'tr'
    }
  } catch (e) {
    console.debug('[region] locale failed:', e)
  }
  return 'en'
}

export async function detectPlayer(auth: LocalAuth): Promise<[string, string]> {
  try {
    const r = await auth.localGet('/chat/v1/session')
    if (r.status === 200) return [r.data.game_name || '', r.data.game_tag || '']
  } catch (e) {
    console.debug('[region] player failed:', e)
  }
  return ['', '']
}
